crates/control/src/visual_referee_filter.rs

```
@@ -0,0 +1,78 @@
+
+    pub hardware: HardwareInterface,
+}
+
+impl VisualRefereeFilter {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            last_primary_state: PrimaryState::Unstiff,
+        })
+    }
+
+    pub fn cycle(&mut self, context: CycleContext<impl Interface>) {
+        let send_game_controller_visual_referee_return_message = matches!(
+            (self.last_primary_state, *context.primary_state),
+            (PrimaryState::Set, PrimaryState::Playing)
+                | (
+                    PrimaryState::Playing,
+                    PrimaryState::Finished | PrimaryState::Ready
+                )
+        );
+        self.last_primary_state = *context.primary_state;
+
+        // Initially a random visual referee decision
+        let mut rng = thread_rng();
+        let gesture = VisualRefereeDecision::from_u32(rng.gen_range(1..=13)).unwrap();
+
+        if send_game_controller_visual_referee_return_message {
+            let mut duration_since_last_whistle = context
+                .filtered_whistle
+                .last_detection
+                .map(|last_detection| {
+                    context
+                        .cycle_time
+                        .start_time
+                        .duration_since(last_detection)
+                        .unwrap()
+                })
+                .unwrap_or(Duration::from_secs(15));
+            if duration_since_last_whistle.as_secs_f32() < 1.0 {
+                duration_since_last_whistle = Duration::from_secs(5)
+            }
+            let message = OutgoingMessage::VisualReferee(VisualRefereeMessage {
+                player_number: *context.player_number,
+                gesture,
+                whistle_age: duration_since_last_whistle,
+            });
+            let _ = context
+                .hardware
+                .write_to_network(message)
+                .wrap_err("failed to write VisualRefereeMessage to hardware");
```

Please return the error instead of consuming it into void:

```suggestion
            context
                .hardware
                .write_to_network(message)
                .wrap_err("failed to write VisualRefereeMessage to hardware")?;
```

---

Done